ga4gh/gff3Parser.py

```diff
@@ -0,0 +1,321 @@
+        self.score = score
+        self.strand = strand
+        self.frame = frame
+        self.attributes = copy.deepcopy(attributes)
+        self.gff3Set = gff3Set
+        self.lineNumber = lineNumber
+        self.parents = set()
+        self.children = set()
+
+    @staticmethod
+    def __dotIfNone(val):
+        return val if val is not None else '.'
+
+    def __attributeStr(self, name):
+        """
+        Return name=value for a single attribute
+        """
+        return "{}={}".format(
+            _encodeAttr(name),
+            ",".join([_encodeAttr(v) for v in self.attributes[name]]))
+
+    def __attributeStrs(self):
+        """
+        Return name=value, semi-colon-separated string for attributes,
+        including url-style quoting
+        """
+        return ";".join([self.__attributeStr(name)
+                         for name in self.attributes.iterkeys()])
+
+    def __str__(self):
+        """
+        Return the object as a valid GFF3 record line.
+        """
+        return "\t".join([self.seqname, self.source, self.type,
+                          str(self.start), str(self.end),
+                          self.__dotIfNone(self.score),
+                          self.__dotIfNone(self.strand),
+                          self.__dotIfNone(self.frame),
+                          self.__attributeStrs()])
+
+    @property
+    def featureId(self):
+        """
+        ID attribute or None if records doesn't have an ID
+        """
+        featId = self.attributes.get("ID")
+        if featId is not None:
+            featId = featId[0]
+        return featId
+
+    def getRequiredAttribute(self, name):
+        "get a require attributes list of values, error if not found"
+        values = self.attributes.get(name)
+        if values is None:
+            raise GFF3Exception(
+                "required attribute not found: {}".format(name),
+                self.gff3Set.fileName, self.lineNumber)
+        return values
+
+
+class Gff3Set(object):
+    """
+    A set of GFF3 sequence annotations
+    """
+    def __init__(self, fileName=None):
+        self.fileName = fileName
+        self.roots = set()     # root nodes (those with out parents)
+        # index of features by id. GFF3 allows disjoint features with
+        # the same id.  None is used to store features without ids
+        self.byFeatureId = collections.defaultdict(list)
+
+    def add(self, feature):
+        """
+        Add a feature record
+        """
+        # None featureId allowed
+        self.byFeatureId[feature.featureId].append(feature)
+
+    def __linkFeature(self, feature):
+        """
+        Link a feature with it's parents.
+        """
+        parentIds = feature.attributes.get("Parent")
+        if parentIds is None:
+            self.roots.add(feature)
+        else:
+            for parentId in parentIds:
+                self.__linkToParent(feature, parentId)
+
+    def __linkToParent(self, feature, parentId):
```

Comment:
The rationale for ``_`` is 95% data hiding, but also leaving the door open to inheritance in the odd occasions that it turns out to be useful. I used to use ``__`` very systematically in the past but I ended up finding it causing quite a few annoying issues (stuff like, even if a particular method is public, if it uses private variables it will fail in obscure ways), and basically requiring a total replumbing of a class if it turned out I wanted to create a subclass. So, I decided that protected variables were generally more useful and just went with those in the majority of cases.

I don't think there's anything wrong with inheritance, it just has to be used judiciously like any other programming technique. 

As I say, the motivation is almost entirely data hiding by convention, and discouraging the access of anything but simple read-only attributes via ``object.attribute``. I think making attributes public by default leads to all sorts of horrible interdependencies between objects, with object X reaching in and modifying attributes of object Y, making it very hard to figure out where code actually resides. By keeping attributes protected by default we discourage this sort of practice without being overly stringent.

Re Python 2 and 3, there's not all that much to look out for, as we've covered most of the truly nasty gotchas by using the futures imports. I think I've had to make about 5 to 10 comments during the entire history of the project asking people to stop using Python 2 only constructs.